### PR TITLE
Allow for changing TZ vairable and cache it for Local timezone

### PR DESCRIPTION
**Pull request by esheppa**
This should solve https://github.com/chronotope/chrono/issues/834.

As discussed, I've done some basic benchmarks, and it seems that this has minimal effect on the standard `/etc/localtime` case, while being around 3% more expensive in the TZ variable case. However this seems justified as this now allows for the TZ variable to change.

### empty TZ variable, timezone is a symlink at `/etc/localtime`

Before (e0b16e9c…):
```
bench_get_local_time    time:   [79.480 ns 79.692 ns 79.923 ns]
```

After (360478ce…):
```
bench_get_local_time    time:   [79.232 ns 79.383 ns 79.556 ns]
```

### with TZ variable

Before (e0b16e9c…):
```
bench_get_local_time    time:   [77.322 ns 77.664 ns 78.002 ns]
```

After (360478ce…):
```
bench_get_local_time    time:   [79.711 ns 79.921 ns 80.155 ns]
```

### CODE REVIEW

File: src/offset/local/tz_info/timezone.rs [hunk 1]
```diff
@@ -26,9 +27,9 @@ impl TimeZone {
     ///
     /// This method in not supported on non-UNIX platforms, and returns the UTC time zone instead.
     ///
-    pub(crate) fn local() -> Result<Self, Error> {
-        if let Ok(tz) = std::env::var("TZ") {
-            Self::from_posix_tz(&tz)
+    pub(crate) fn local(env_result: &Result<String, VarError>) -> Result<Self, Error> {
```
Comment (esheppa) [inline src/offset/local/tz_info/timezone.rs:30]: this may be better as `&Option<String>` or similar

File: src/offset/local/tz_info/timezone.rs [hunk 2]
```diff
@@ -7,6 +7,7 @@ use std::{cmp::Ordering, fmt, str};
 
 use super::rule::{AlternateTime, TransitionRule};
 use super::{parser, Error, DAYS_PER_WEEK, SECONDS_PER_DAY};
+use std::env::VarError;
```
Comment (djc) [inline src/offset/local/tz_info/timezone.rs:10]: This should go in the `std` block.

File: src/offset/local/tz_info/timezone.rs (hunk 1 above)
Comment (djc) [reply src/offset/local/tz_info/timezone.rs:30]: So this is in order to reuse the result for both `Source::new()` and `TimeZone::local()`, right? I think we should probably pass it as `Option<&str>`. Also this change might make sense to do as a separate commit?

File: src/offset/local/unix.rs [hunk 3]
```diff
@@ -31,69 +31,70 @@ thread_local! {
     static TZ_INFO: RefCell<Option<Cache>> = Default::default();
 }
 
+// weird looking function signature
+// but allows sharing the results of env::var around
+fn get_tz_env_hash(env_result: &Result<String, env::VarError>) -> Option<u64> {
```
Comment (djc) [inline src/offset/local/unix.rs:36]: Given that this is single-caller, not sure why extracting it as a separate function is useful.

File: src/offset/local/unix.rs [hunk 4]
```diff
@@ -31,69 +31,70 @@ thread_local! {
...
-                Ok(_) | Err(_) => *mtime,
-            },
-            Source::Environment => return false,
-        };
-
-        match Source::default() {
-            Source::LocalTime { mtime, .. } => {
-                *self = Source::LocalTime { mtime, last_checked: now };
-                prev != mtime
+    fn out_of_date(&self, other_source: &Source) -> bool {
```
Comment (djc) [inline src/offset/local/unix.rs:71]: Add an empty line before this? I'd also rename `other_source` as `prev` or something, since you're already using names like `old_mtime` and `old_hash` for some of the inner data it would make sense to use the more specific semantics here.

File: src/offset/local/unix.rs [hunk 5]
```diff
@@ -115,17 +116,41 @@ fn fallback_timezone() -> Option<TimeZone> {
...
     }
 }
 
 impl Cache {
     fn offset(&mut self, d: NaiveDateTime, local: bool) -> LocalResult<DateTime<Local>> {
-        if self.source.out_of_date() {
-            *self = Cache::default();
+        let now = SystemTime::now();
+
+        match now.duration_since(self.last_checked) {
```
Comment (djc) [inline src/offset/local/unix.rs:135]: So this block has been moved from `Source::out_of_date()` to here, can we do that as a separate commit?

**Review by djc — COMMENTED** [review]
Good idea to add a benchmark! I think these are good changes, asking for a bit more structure in the commit history so it becomes easier to infer the rationale for the constituent parts.

File: src/offset/local/tz_info/timezone.rs (hunk 1 above)
Comment (esheppa) [reply src/offset/local/tz_info/timezone.rs]: excellent, I've gone for `Option<&str>` in [d2a32c8](https://github.com/chronotope/chrono/pull/853/commits/d2a32c8e50a44fdfd5f33bc4fcc11b12da5507eb)

File: src/offset/local/unix.rs (hunk 3 above)
Comment (esheppa) [reply src/offset/local/unix.rs]: removed

File: src/offset/local/unix.rs (hunk 4 above)
Comment (esheppa) [reply src/offset/local/unix.rs]: I've ended up in-lining this as above, which resulted in a nicer progression from the old code to the new code

File: src/offset/local/unix.rs (hunk 5 above)
Comment (esheppa) [reply src/offset/local/unix.rs:108]: this is now done in [e46e357](https://github.com/chronotope/chrono/pull/853/commits/e46e357ca5c623652d3b71fa7da865ac586ff184)

File: src/offset/local/unix.rs [hunk 6]
```diff
@@ -32,68 +32,40 @@ thread_local! {
...
-    // we don't bother storing the contents of the environment variable in this case.
-    // changing the environment while the process is running is generally not reccomended
-    Environment,
+    LocalTime { mtime: SystemTime },
+    Environment { hash: u64 },
 }
 
-impl Default for Source {
-    fn default() -> Source {
-        // use of var_os avoids allocating, which is nice
```
Comment (esheppa) [inline src/offset/local/unix.rs:43]: I benchmarked using `var_os` vs `var` and the difference was negligible, while the code was messier, hence the removal of this comment and use of `var_os`

File: src/offset/local/unix.rs [hunk 7]
```diff
@@ -89,14 +89,20 @@ fn fallback_timezone() -> Option<TimeZone> {
...
-            zone: TimeZone::local().ok().or_else(fallback_timezone).unwrap_or_else(TimeZone::utc),
-            source: Source::default(),
             last_checked: SystemTime::now(),
+            source: Source::new(env_ref),
+            zone: new_zone(env_ref),
         }
     }
 }
 
+fn new_zone(var: Option<&str>) -> TimeZone {
```
Comment (djc) [inline src/offset/local/unix.rs:100]: I find the `new_` prefix a little confusing here. In the end this is just the current zone, right? Maybe just call this `current()` or `zone()` or `current_zone()`?

**Review by djc — APPROVED** [review]

File: src/offset/local/unix.rs (hunk 7 above)
Comment (esheppa) [reply src/offset/local/unix.rs]: good call, I've renamed to `current_zone`

**Review by djc — APPROVED** [review]
